### PR TITLE
EZP-29115: Add characters counter to RichText

### PR DESCRIPTION
**Pull request by konradoboza**
| Question      | Answer
| ------------- | ---
| Tickets       | https://jira.ez.no/browse/EZP-29115
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Tests pass?   | yes
| Doc needed?   | yes
| License       | [GPL-2.0](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/LICENSE)

This PR contains a proof of concept for one of the features which was requested by the customer. This is a simple characters counter which is enabled/disabled globally for RichText via User Preferences. The idea is to count all characters inside tags except:
- custom tags,
- embeds,
- images.

The matter of styling and configuration could be changed, it is rather a basic idea which in my opinion seems to be consistent with the rest of the back office. However all feedback will be warmly welcomed. :) Below I attached some screenshot which presents current state of the feature:

counter

user_settings

changing_visibility

User settings were moved to `ezplatform-user` repository, so it is dependant on: https://github.com/ezsystems/ezplatform-user/pull/8
#### Checklist:
- [x] Coding standards (`$ composer fix-cs`)
- [x] Ready for Code Review
- [ ] nested custom tags handling

### CODE REVIEW

**Comment by konradoboza** [comment]
User settings part was extracted to `ezplatform-user` repository, ref: https://github.com/ezsystems/ezplatform-user/pull/8.

Implementation was moved from the plugin to `base-rich-text.js` as suggested by @dew326. Counting characters inside custom tags need to be improved.

File: src/bundle/Resources/public/css/alloyeditor/alloyeditor-ez.css [hunk 1]
```diff
@@ -1058,3 +1058,18 @@ pre.ez-has-anchor > .ez-icon {
     font-size: 14px;
     line-height: 1;
 }
+
+.ez-character-counter {
+    position: relative;
+    width: 99.8%;
```
Comment (dew326) [inline src/bundle/Resources/public/css/alloyeditor/alloyeditor-ez.css]: why?

File: src/bundle/Resources/public/css/alloyeditor/alloyeditor-ez.css [hunk 2]
```diff
@@ -1058,3 +1058,18 @@ pre.ez-has-anchor > .ez-icon {
...
+    color: #696969;
```
Comment (dew326) [inline src/bundle/Resources/public/css/alloyeditor/alloyeditor-ez.css]: Please use colors from our custom.scss variables

File: src/bundle/Resources/public/css/alloyeditor/alloyeditor-ez.css [hunk 3]
```diff
@@ -1058,3 +1058,18 @@ pre.ez-has-anchor > .ez-icon {
...
+    background-color: #EDEDED;
```
Comment (dew326) [inline src/bundle/Resources/public/css/alloyeditor/alloyeditor-ez.css]: Please use colors from our custom.scss variables

File: src/bundle/Resources/public/css/alloyeditor/alloyeditor-ez.css [hunk 4]
```diff
@@ -1058,3 +1058,18 @@ pre.ez-has-anchor > .ez-icon {
...
+    font-size: 0.7rem;
```
Comment (dew326) [inline src/bundle/Resources/public/css/alloyeditor/alloyeditor-ez.css]: Please use `calculcateRem()`, applies to all `Xrem`

File: src/bundle/Resources/public/css/alloyeditor/alloyeditor-ez.css [hunk 5]
```diff
@@ -1058,3 +1058,18 @@ pre.ez-has-anchor > .ez-icon {
...
+    margin: -1.54em auto 0 auto;
+    border-bottom-left-radius: 0.25rem;
+    border-bottom-right-radius: 0.25rem;
+}
+
+.ez-character-counter span {
```
Comment (dew326) [inline src/bundle/Resources/public/css/alloyeditor/alloyeditor-ez.css]: Please add some class to this `span`

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 6]
```diff
@@ -220,6 +222,44 @@
 
             return alloyEditor;
         }
+
+        countWordsCharacters(container, editorHtml) {
+            let counterWrapper = container.parentElement.getElementsByClassName('ez-character-counter');
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: `querySelector('.ez-character-counter')` and `const` not `let`

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 7]
```diff
@@ -220,6 +222,44 @@
...
+
+            if (counterWrapper.length) {
+                let wordCharacterWrappers = counterWrapper[0].getElementsByTagName('span');
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: `const`, please add some class to this span and do `querySelector`

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 8]
```diff
@@ -220,6 +222,44 @@
...
+                let countableTags = this.fetchCountableTags(editorHtml);
+
+                countableTags.forEach((tag) => {
+                    let sanitizedText = this.cleanWhiteCharacters(tag.innerText);
+
+                    wordCount += sanitizedText ? sanitizedText.split(' ').length : 0;
+                    characterCount += sanitizedText.length;
+                });
+
+                wordCharacterWrappers[0].innerText = wordCount;
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: You can't rely on the order of getting element. What if someone will add another span before yours?

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 9]
```diff
@@ -220,6 +222,44 @@
...
         }
+
+        countWordsCharacters(container, editorHtml) {
+            let counterWrapper = container.parentElement.getElementsByClassName('ez-character-counter');
+
+            if (counterWrapper.length) {
+                let wordCharacterWrappers = counterWrapper[0].getElementsByTagName('span');
+                let characterCount = 0;
+                let wordCount = 0;
+                let countableTags = this.fetchCountableTags(editorHtml);
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: `const`

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 10]
```diff
@@ -196,6 +196,8 @@
                     this.xhtmlNamespace,
                     this.ezNamespace
                 );
+
+                this.countWordsCharacters(container, data);
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: You should pass `doc` not `data`. `doc` is cleaned from unnecessary items.

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 11]
```diff
@@ -220,6 +222,44 @@
...
+            let counterWrapper = container.parentElement.getElementsByClassName('ez-character-counter');
+
+            if (counterWrapper.length) {
+                let wordCharacterWrappers = counterWrapper[0].getElementsByTagName('span');
+                let characterCount = 0;
+                let wordCount = 0;
+                let countableTags = this.fetchCountableTags(editorHtml);
+
+                countableTags.forEach((tag) => {
+                    let sanitizedText = this.cleanWhiteCharacters(tag.innerText);
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: `const`

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 12]
```diff
@@ -220,6 +222,44 @@
...
+                wordCharacterWrappers[0].innerText = wordCount;
+                wordCharacterWrappers[1].innerText = characterCount;
+            }
+        }
+
+        fetchCountableTags(html) {
+            const allowedTags = ['p:not(.ez-embed-content)', 'li', 'h1', 'h2', 'h3', 'h4', 'h5' , 'h6', 'th', 'td'];
+            const notCustomTagSelector = ':not([data-ezelement=ezattributes]) > ';
+
+            let allowedSelectors = allowedTags.map((item) => {
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: `const`

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 13]
```diff
@@ -220,6 +222,44 @@
...
+
+                wordCharacterWrappers[0].innerText = wordCount;
+                wordCharacterWrappers[1].innerText = characterCount;
+            }
+        }
+
+        fetchCountableTags(html) {
+            const allowedTags = ['p:not(.ez-embed-content)', 'li', 'h1', 'h2', 'h3', 'h4', 'h5' , 'h6', 'th', 'td'];
+            const notCustomTagSelector = ':not([data-ezelement=ezattributes]) > ';
+
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: not needed

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 14]
```diff
@@ -220,6 +222,44 @@
...
+            let allowedSelectors = allowedTags.map((item) => {
+                return notCustomTagSelector.concat(item);
+            });
+
+            return new DOMParser()
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: If you will pass `doc` you don't need to do this parser. You can do querySelector on `doc`

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 15]
```diff
@@ -220,6 +222,44 @@
...
+                    wordCount += sanitizedText ? sanitizedText.split(' ').length : 0;
+                    characterCount += sanitizedText.length;
+                });
+
+                wordCharacterWrappers[0].innerText = wordCount;
+                wordCharacterWrappers[1].innerText = characterCount;
+            }
+        }
+
+        fetchCountableTags(html) {
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: `getCountableTags`

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js (hunk 9 above)
Comment (konradoboza) [reply src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: `characterCount` and `wordCount` are modified right after, they should also be set as `const`?

**Comment by konradoboza** [comment]
@dew326 remarks applied, please take a look.

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js (hunk 9 above)
Comment (dew326) [reply src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: No, you will not be able to assign a new value to these variables.

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 16]
```diff
@@ -220,6 +222,43 @@
 
             return alloyEditor;
         }
+
+        countWordsCharacters(container, editorHtml) {
+            const counterWrapper = container.parentElement.querySelector('.ez-character-counter');
+
+            if (counterWrapper) {
+                const countableTags = this.getCountableTags(editorHtml);
+                const wordWrapper = counterWrapper.querySelector('span.ez-character-counter__word-count');
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: `querySelector('.ez-character-counter__word-count')` is enough

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 17]
```diff
@@ -220,6 +222,43 @@
...
+                const charactersWrapper = counterWrapper.querySelector('span.ez-character-counter__character-count');
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: like above.

File: src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js [hunk 18]
```diff
@@ -220,6 +222,43 @@
...
+
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/fieldType/base/base-rich-text.js]: No empty line between `const` and `let`

File: src/bundle/Resources/public/scss/alloyeditor.scss [hunk 19]
```diff
@@ -7,3 +7,4 @@
 @import 'alloyeditor-link-edit';
 @import 'alloyeditor-custom-tag';
 @import 'alloyeditor-custom-styles';
+@import 'alloyeditor-character-counter.scss';
```
Comment (dew326) [inline src/bundle/Resources/public/scss/alloyeditor.scss]: `'alloyeditor-character-counter'` is enough

File: src/bundle/Resources/public/scss/_alloyeditor-character-counter.scss [hunk 20]
```diff
@@ -0,0 +1,15 @@
+.ez-character-counter {
+  position: relative;
```
Comment (dew326) [inline src/bundle/Resources/public/scss/_alloyeditor-character-counter.scss]: indentation 4 spaces.

**Comment by konradoboza** [comment]
ping @adamwojs @dew326 @tischsoic

File: src/bundle/Resources/translations/user_settings.en.xliff [hunk 21]
```diff
@@ -61,6 +61,36 @@
         <target state="new">Sub-items</target>
         <note>key: settings.subitems_limit.value.title</note>
       </trans-unit>
+      <trans-unit id="98400a7e546231a994f6fea03dce3ebd52c1c33d" resname="settings.character_counter.value.description">
```
Comment (adamwojs) [inline src/bundle/Resources/translations/user_settings.en.xliff]: Note to my self: This file should be moved to `ezplatform-user`.

**Pull request merged**